kong/plugins/proxy-cache/handler.lua

```diff
@@ -9,7 +9,6 @@ local floor            = math.floor
 local get_method       = ngx.req.get_method
 local ngx_get_uri_args = ngx.req.get_uri_args
 local ngx_get_headers  = ngx.req.get_headers
-local resp_get_headers = ngx.resp.get_headers
```

Comment:
So long as `resp_get_headers()` does not get invoked in the init phase, it shouldn't be a problem to require it here. This fix doesn't seem necessary at a first glance, unless I missed something. Maybe a stacktrace would be helpful?
```
error_log logs/error.log error;

events {}

http {
    init_by_lua_block {
        require "resty.core"
        local resp_get_headers = ngx.resp.get_headers
        print("OK")
    }
}
```

`logs/error.log`:
```
2019/05/20 13:00:01 [notice] 25952#0: [lua] init_by_lua:4: OK
```

Reply:
I can't reproduce the problem in an isolated conf either, but it happens when `load_module_if_exists()` from `kong.tools.utils` `require` the plugin :
```
spec/02-integration/02-cmd/02-start_stop_spec.lua:46: Error: ./kong/cmd/start.lua:64: nginx: [error] init_by_lua error:
 ...c537d39b3000/luarocks/share/lua/5.1/kong/tools/utils.lua:590: ...rocks/share/lua/5.1/kong/plugins/proxy-cache/handler.lua:12: attempt to index field 'resp' (a nil value)
stack traceback:
	[C]: in function 'error'
	...c537d39b3000/luarocks/share/lua/5.1/kong/tools/utils.lua:590: in function 'load_module_if_exists'
	...7d39b3000/luarocks/share/lua/5.1/kong/db/dao/plugins.lua:151: in function 'load_plugin_handler'
	...7d39b3000/luarocks/share/lua/5.1/kong/db/dao/plugins.lua:223: in function 'load_plugin'
	...7d39b3000/luarocks/share/lua/5.1/kong/db/dao/plugins.lua:270: in function 'load_plugin_schemas'
	...af9e436c537d39b3000/luarocks/share/lua/5.1/kong/init.lua:377: in function 'init'
	init_by_lua:16: in main chunk
```

Reply:
Are you using a particularly old version of OpenResty in the environment that produces this error?

Reply:
It is OpenResty 1.13.6.2.

Reply:
Went to the bottom of this: this particular test fails because it configures a stream listener, which instantiates a stream Lua VM. In this implementation (ngx_stream_lua), `ngx.resp` is not defined. Loading this plugin works fine in the http submodule, but this is the issue:

```nginx
error_log /dev/stderr error;

events {}

stream {
    init_by_lua_block {
        require "resty.core"
        local resp_get_headers = ngx.resp.get_headers
        print("OK")
    }
}
```

```
$ nginx -p .         
nginx: [error] init_by_lua error: init_by_lua:3: attempt to index field 'resp' (a nil value)
stack traceback:
        init_by_lua:3: in main chunk
```

Until plugins can be explicitly enabled/disabled on http/stream submodules (proxy-cache does not make sense in the stream submodule), this is a more appropriate fix that does not affect the http submodule and still allows the stream Lua VM to be initialized:

```diff
diff --git a/kong/plugins/proxy-cache/handler.lua b/kong/plugins/proxy-cache/handler.lua
index 8ace2c1..22cb892 100644
--- a/kong/plugins/proxy-cache/handler.lua
+++ b/kong/plugins/proxy-cache/handler.lua
@@ -9,7 +9,7 @@ local floor            = math.floor
 local get_method       = ngx.req.get_method
 local ngx_get_uri_args = ngx.req.get_uri_args
 local ngx_get_headers  = ngx.req.get_headers
-local resp_get_headers = ngx.resp.get_headers
+local resp_get_headers = ngx.resp and ngx.resp.get_headers
 local ngx_log          = ngx.log
 local ngx_now          = ngx.now
 local ngx_re_gmatch    = ngx.re.gmatch
```

Reply:
Great, thank you! I made the changes according to your suggestion.